core/local/local_test.go

```diff
@@ -96,6 +101,480 @@ func TestExecutionSchedulerRun(t *testing.T) {
+							errors.add(1);
+						}
+						socket.close()
+					});
+					socket.on('error', function (e) {
+						console.log('ws error: ' + e.error());
+						errors.add(1);
+					});
+				}
+			);
+			check(response, { 'status is 101': (r) => r && r.status === 101 }, customTags);
+		});
+	}
+`)
+
+	runner, err := js.New(&loader.SourceData{
+		URL:  &url.URL{Path: "/script.js"},
+		Data: []byte(script)},
+		nil, lib.RuntimeOptions{Env: map[string]string{"TESTGLOBALVAR": "global"}})
+	require.NoError(t, err)
+
+	// Required for WS tests
+	opts := runner.GetOptions()
+	opts.Hosts = map[string]net.IP{
+		"httpbin.local": net.ParseIP("127.0.0.1"),
+	}
+	require.NoError(t, runner.SetOptions(opts))
+
+	logger := logrus.New()
+	logger.SetOutput(testutils.NewTestOutput(t))
+	execScheduler, err := NewExecutionScheduler(runner, logger)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	done := make(chan struct{})
+	samples := make(chan stats.SampleContainer)
+	go func() {
+		assert.NoError(t, execScheduler.Init(ctx, samples))
+		assert.NoError(t, execScheduler.Run(ctx, ctx, samples))
+		close(done)
+	}()
+
+	expectedTrailTags := []map[string]string{
+		{"testtag1": "scenario1", "reqtag": "scenario1"},
+		{"testtag2": "scenario2", "reqtag": "scenario2"},
+	}
+	expectedNetTrailTags := []map[string]string{
+		{"testtag1": "scenario1"},
+		{"testtag2": "scenario2"},
+	}
+	expectedConnSampleTags := map[string]string{
+		"testtag3": "scenario3", "wstag": "scenario3",
+	}
+	expectedPlainSampleTags := []map[string]string{
+		{"testtag3": "scenario3"},
+		{"testtag3": "scenario3", "wstag": "scenario3"},
+	}
+	var gotSampleTags int
+	for {
+		select {
+		case sample := <-samples:
+			if s, ok := sample.(stats.Sample); ok && s.Metric.Name == "errors" {
+				assert.FailNow(t, "received error sample from test")
+			}
+			if trail, ok := sample.(*httpext.Trail); ok {
+				tags := trail.Tags.CloneTags()
+				for _, expTags := range expectedTrailTags {
+					if reflect.DeepEqual(expTags, tags) {
+						gotSampleTags++
+					}
+				}
+			}
+			if netTrail, ok := sample.(*netext.NetTrail); ok {
+				tags := netTrail.Tags.CloneTags()
+				for _, expTags := range expectedNetTrailTags {
+					if reflect.DeepEqual(expTags, tags) {
+						gotSampleTags++
+					}
+				}
+			}
+			if cs, ok := sample.(stats.ConnectedSamples); ok {
+				for _, s := range cs.Samples {
+					tags := s.Tags.CloneTags()
+					if reflect.DeepEqual(expectedConnSampleTags, tags) {
+						gotSampleTags++
+					}
+				}
+			}
+			if s, ok := sample.(stats.Sample); ok &&
+				(s.Metric.Name == "checks" || s.Metric.Name == "group_duration") {
+				tags := s.Tags.CloneTags()
+				for _, expTags := range expectedPlainSampleTags {
+					if reflect.DeepEqual(expTags, tags) {
+						gotSampleTags++
+					}
+				}
+			}
+		case <-done:
+			require.Equal(t, 8, gotSampleTags, "received wrong amount of samples with expected tags")
```

Comment:
So a couple of issues with this test:
1. It's too large, and possibly testing too much functionality. I'd be up for splitting it, but I think we'd end up with more LOCs overall. Any ideas on simplifying, or splitting criteria and/or where to move it?
2. This assertion is far too brittle (i.e. easily susceptible to changes), and the counter approach is not very reliable. Any better ideas? :)